audinterface/core/segment_with_feature.py

```diff
@@ -0,0 +1,825 @@
+        index = audformat.segmented_index(files, starts, ends)
+        if len(index) == 0:
+            # Pass no data to ensure consistent dtype for columns
+            result = pd.DataFrame(
+                index=audformat.segmented_index(), columns=self.feature_names
+            )
+        else:
+            result = pd.DataFrame(
+                index=index,
+                data=features,
+                columns=self.feature_names,
+            )
+
+        dtypes = [table[col].dtype for col in table.columns]
+        labels = {
+            col: pd.Series(
+                labels[:, ncol], index=index, dtype=dtypes[ncol]
+            )  # supports also category
+            for ncol, col in enumerate(table.columns)
+        }
+        table = pd.DataFrame(labels, index, columns=table.columns)
+
+        result = result.join(table)
+        return result
+
+    def _check_return_format(self, x):
+        if not isinstance(x, pd.Series):
+            raise ValueError(f"A series must be returned, but got {type(x)}")
+        utils.assert_index(x.index)
+        # Disallow filewise index
+        if audformat.is_filewise_index(x.index):
+            raise ValueError(
+                "The returned series must have a signal index or a segmented index"
+            )
+
+    def _construct_frame_segmented_index(self, y: pd.Series):
+        r"""Construct dataframe with segmented index from process result."""
+        files = []
+        starts = []
+        ends = []
+        features = {col: [] for col in self.feature_names}
+        for (file, start, _), series in y.items():
+            self._check_return_format(series)
+            df = self._series_to_frame(series)
+            files.extend([file] * len(df))
+            starts.extend(df.index.get_level_values("start") + start)
+            ends.extend(df.index.get_level_values("end") + start)
+            for col in self.feature_names:
+                features[col].extend(df[col])
+        if not files:
+            # Pass no data to ensure consistent dtype for columns
+            return pd.DataFrame(
+                index=audformat.segmented_index(), columns=self.feature_names
+            )
+        return pd.DataFrame(
+            index=audformat.segmented_index(files, starts, ends),
+            data=features,
+            columns=self.feature_names,
+        )
+
+    def _construct_frame_signal_index(self, y: pd.Series):
+        r"""Construct dataframe with signal index from process result."""
+        starts = []
+        ends = []
+        features = {col: [] for col in self.feature_names}
+        for (start, _), series in y.items():
+            self._check_return_format(series)
+            df = self._series_to_frame(series)
+            starts.extend(df.index.get_level_values("start") + start)
+            ends.extend(df.index.get_level_values("end") + start)
+            for col in self.feature_names:
+                features[col].extend(df[col])
```

Comment:
Maybe you can try to avoid creating additional dataframes and calling `get_level_values()` by something similar to this:

```suggestion
        for (start, _), series in y.items():
            self._check_return_format(series)
            if series.empty:
                data = np.empty((1, len(self.feature_names)))
            else:
                data = [self._reshape_numpy_1d(values) for values in series]
                data = np.stack(data).T
            starts.extend(series.index.levels[0] + start)
            ends.extend(series.index.levels[1] + start)
            for col, val in zip(self.feature_names, data):
                features[col].extend(val)
```

I have not tested if this is really faster, but I would assume it should be.

Reply:
The following code should work instead:

```suggestion
        for (start, _), series in y.items():
            self._check_return_format(series)
            if series.empty:
                data = np.empty((1, len(self.feature_names)))
            else:
                data = [self._reshape_numpy_1d(values) for values in series]
                data = np.stack(data).T
            index_data = series.index.to_numpy()
            if len(index_data) > 0:
                starts.extend([t[0] + start for t in index_data])
                ends.extend([t[1] + start for t in index_data])
            for col, val in zip(self.feature_names, data):
                features[col].extend(val)
```

I updated `_construct_frame_segmented_index()` in the same way, and just got one failing test:

```
E       AssertionError: Attributes of DataFrame.iloc[:, 0] (column name="mean") are different
E       
E       Attribute "dtype" are different
E       [left]:  float32
E       [right]: float64
```

I then first rerun your benchmark with my changes:

```
audinterface.Feature: Average elapsed time over 100 repetitions: 0.385 seconds                      
audinterface.SegmentWithFeature: Average elapsed time over 100 repetitions: 0.744 seconds
```

and repeated it with my changes applied:

```
audinterface.Feature: Average elapsed time over 100 repetitions: 0.378 seconds                      
audinterface.SegmentWithFeature: Average elapsed time over 100 repetitions: 0.692 seconds
```

Unfortunately, this does not really help much.

Reply:
Maybe, we just stay with your proposed implementation?

Reply:
I integrated the suggestion, and adjusted the tests. The failing test revealed that the returned dtype wasn't kept when constructing the dataframe before, and that the suggestion was more correct than the previous implementation :)

I adjusted and added some tests to check for the correct handling of float dtypes, and added the helper `_reshape_numpy()` to cover the case that the returned series contains scalar values, so we still preserve the dtype in that case.

I also reran the benchmark script again, and there is a similar improvement in execution time to before.